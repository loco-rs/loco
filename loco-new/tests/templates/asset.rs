use loco::{settings, wizard::AssetsOption};
use rstest::rstest;

use super::*;
use crate::assertion;

pub fn run_generator(asset: AssetsOption) -> TestGenerator {
    let settings = settings::Settings {
        asset: asset.into(),
        ..Default::default()
    };

    TestGenerator::generate(settings)
}

#[rstest]
fn test_config_file_middleware_when_asset_empty(
    #[values("config/development.yaml", "config/test.yaml")] config_file: &str,
) {
    let generator: TestGenerator = run_generator(AssetsOption::None);
    let content = assertion::yaml::load(generator.path(config_file));

    assertion::yaml::assert_path_is_empty(&content, &["server", "middlewares"]);
}

#[rstest]
fn test_config_file_middleware_asset_server(
    #[values("config/development.yaml", "config/test.yaml")] config_file: &str,
) {
    let generator: TestGenerator = run_generator(AssetsOption::Serverside);
    let content = assertion::yaml::load(generator.path(config_file));

    assertion::yaml::assert_path_is_object(&content, &["server", "middlewares", "static"]);

    let expected: serde_yaml::Value = serde_yaml::from_str(
        r"
enable: true
must_exist: true
precompressed: false
folder:
    uri: /static
    path: assets/static
fallback: assets/static/404.html
",
    )
    .unwrap();
    assertion::yaml::assert_path_value_eq(
        &content,
        &["server", "middlewares", "static"],
        &expected,
    );
}

#[rstest]
fn test_config_file_middleware_asset_client(
    #[values("config/development.yaml", "config/test.yaml")] config_file: &str,
) {
    let generator: TestGenerator = run_generator(AssetsOption::Clientside);
    let content = assertion::yaml::load(generator.path(config_file));

    assertion::yaml::assert_path_is_object(&content, &["server", "middlewares", "static"]);

    let expected: serde_yaml::Value = serde_yaml::from_str(
        r"
enable: true
must_exist: true
precompressed: false
folder:
    uri: /
    path: frontend/dist
fallback: frontend/dist/index.html
",
    )
    .unwrap();
    assertion::yaml::assert_path_value_eq(
        &content,
        &["server", "middlewares", "static"],
        &expected,
    );
}

#[rstest]
fn test_cargo_toml(
    #[values(AssetsOption::None, AssetsOption::Serverside, AssetsOption::Clientside)]
    asset: AssetsOption,
) {
    let generator = run_generator(asset.clone());
    let content = assertion::toml::load(generator.path("Cargo.toml"));

    insta::assert_snapshot!(
        format!("cargo_dependencies_{:?}", asset),
        content.get("dependencies").unwrap()
    );
}
