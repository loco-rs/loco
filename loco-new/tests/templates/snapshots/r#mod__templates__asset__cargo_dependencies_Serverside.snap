---
source: loco-new/tests/templates/asset.rs
expression: "content.get(\"dependencies\").unwrap()"
---
{ async-trait = "0.1.74", axum = "0.7.5", serde_json = "1", tracing = "0.1.40", unic-langid = "0.9.4", fluent-templates = { features = ["tera"], version = "0.8.0" }, loco-rs = { workspace = true }, serde = { features = ["derive"], version = "1" }, tokio = { default-features = false, features = ["rt-multi-thread"], version = "1.33.0" }, tracing-subscriber = { features = ["env-filter", "json"], version = "0.3.17" } }
